3단계 - 로또(2등)

ghtjr410 opened the pull request:
# 로또(2등) 과제 제출
안녕하세요!
로또(2등) 과제 제출합니다.
## 미션 진행 과정
스트랭글러 패턴을 적용하여 점진적으로 리팩터링했습니다.
1. `LottoRank`에 `fromV2(matchCount, bonusMatch)` 메서드 추가
2. `WinningLotto`에 보너스 번호 개념 도입
3. `WinningLotto.match()`에서 `fromV2` 호출로 전환
4. 기존 `from()` 제거 및 `fromV2` → `from`으로 일원화

실무에서도 변경사항이 많을 때 스트랭글러 패턴을 자주 사용하는데, `이중화 → 전환 → 일원화` 순서로 커밋하면 변경 흐름을 파악하기 좋다고 생각했습니다.
## 고민했던 지점
LottoResult: **record** + 불변 vs **class** + 가변
기존에는 `record`로 불변 객체를 유지했지만, 집계 책임이 호출부에 분산되는 문제가 있었습니다.
``` java
// 기존: 밖에서 수집 → 전달
List<LottoRank> ranks = new ArrayList<>();
for (Lotto lotto : values) {
    ranks.add(winningLotto.match(lotto));
}
return new LottoResult(ranks);
```
``` java
// 변경: 객체가 스스로 집계
LottoResult result = new LottoResult();
for (Lotto lotto : values) {
    result.updateRank(winningLotto.match(lotto));
}
return result;
```
불변성보다 책임 응집을 우선하여 `class + updateRank` 방식을 선택했습니다.
***
리뷰 잘 부탁드립니다!

ghtjr410 committed:
docs: 테스트 품질 체크리스트 문서 업데이트

checklist.md
- 테스트 품질 섹션에 테스트 객체 생성 시 부생성자 도입 검토 항목 추가

ghtjr410 committed:
refactor: Lotto 생성 로직 단순화 및 불필요한 검증 메서드 제거

Lotto.java
- 중복 검증용 toValidatedSet 메서드 제거 및 Set.copyOf로 단순화
- 생성자 로직 간소화
- validateNotEmpty 파라미터 타입을 Set 기반으로 변경

ghtjr410 committed:
refactor: LottoResult 집계 방식 단순화 및 관련 코드 정리

LottoResult.java
- record 제거 후 상태 변경 가능한 클래스 구조로 변경
- EnumMap 기반 누적 집계 방식 도입
- updateRank 메서드로 랭크 처리 일원화

PurchasedLottos.java
- 임시 리스트 제거 후 LottoResult로 직접 누적

LottoResultTest.java
- 제거된 생성자 관련 테스트 삭제

ghtjr410 committed:
refactor: LottoResult 집계 방식 단순화

LottoResult.java
- record 제거 후 EnumMap 기반 집계 방식으로 변경
- updateRank 메서드로 랭크 누적 일원화
- 불필요한 List 생성자 및 검증 제거

PurchasedLottos.java
- 임시 리스트 제거 후 LottoResult로 직접 누적

LottoResultTest.java
- 제거된 생성자 관련 테스트 삭제

ghtjr410 committed:
feat: 로또 등수 보너스 조건 도입 및 등수 체계 개선

LottoRank.java
- 보너스 번호(requireBonus) 여부 추가
- 등수 구성 재정의(SECOND 보너스 적용, THIRD 분리 등)
- fromV2로 일치 갯수 + 보너스 여부 기반 등수 계산 추가
- 기존 from 로직 제거 및 prizeDisplay 정리

LottoRankTest.java
- fromV2 기준으로 테스트 케이스 전면 수정
- 변경된 상금 규칙에 맞춰 prize 계산 테스트 업데이트

LottoResultTest.java
- 변경된 등수(FIFTH) 기준으로 테스트 수정

ghtjr410 committed:
feat: Lotto 생성자 편의 기능 추가 및 테스트 코드 간소화

Lotto.java
- int... 기반 생성자 추가로 LottoNumber 변환 자동화
- 테스트용으로 사용되던 toSet 보조 메서드 정리

LottoTest.java
- createLotto 헬퍼 제거 후 new Lotto(int...) 직접 사용
- 테스트 전반을 int 가변인자 생성자로 단순화

PurchasedLottosTest.java
- 불필요한 헬퍼 메서드 제거
- Lotto(int...) 기반으로 테스트 단순화

ghtjr410 committed:
refactor: Lotto 생성자 내부 중복 제거

Lotto.java
- int... 생성자에서 LottoNumber 변환 로직을 toSet 메서드로 위임하도록 정리

ghtjr410 committed:
feat: 보너스 번호 기능 추가 및 검증 로직 도입

Application.java
- 당첨 번호와 보너스 번호를 각각 입력받도록 변경
- WinningLotto 생성 시 보너스 번호 전달

Lotto.java
- 보너스 번호 중복 체크를 위해 contains 메서드 public으로 변경

LottoNumber.java
- 문자열 기반 생성자 추가로 입력 처리 편의성 향상

WinningLotto.java
- 보너스 번호 필드 추가
- 당첨 번호와 보너스 번호 중복 검증 로직 추가

InputView.java
- 보너스 번호 입력 메서드(readBonusNumber) 추가

WinningLottoTest.java
- 보너스 번호 생성 성공/중복 예외 테스트 추가

ghtjr410 committed:
refactor: 등수 판별 메서드 통합 및 보너스 번호 매칭 적용

LottoRank.java
- fromV2 → from으로 통합
- 일치 개수 + 보너스 여부 기반 등수 판별만 유지
- 기존 단일 파라미터 from 제거

WinningLotto.java
- match 로직에 보너스 번호 일치 여부 반영
- hasBonusMatch 메서드 추가

LottoRankTest.java
- 통합된 from 메서드 기준으로 테스트 수정

ghtjr410 committed:
feat: 등수 표시 문구 추가 및 출력 순서 보완

LottoRank.java
- 각 등수에 displayName 필드 추가
- matchCountDisplay가 displayName을 반환하도록 변경
- ENUM 정의에 사용자 친화적 문구 반영

ResultView.java
- 당첨 결과 출력 시 FIFTH 등수도 포함되도록 출력 순서 보완

ghtjr410 committed:
docs: 보너스 번호 및 등수 체계 변경 반영

03-lotto-rank-second.md
- PR 링크 추가
- LottoNumber 문자열 생성 항목 추가
- WinningLotto 보너스 번호 생성 및 중복 검증 항목 추가
- LottoRank 등수 기준을 보너스 여부 포함 방식으로 업데이트
- 변경된 등수·상금 체계(FIFTH~SECOND)로 체크리스트 수정

javajigi commented on src/main/java/lotto/domain/LottoNumber.java:
## 상황
- 웹 백엔드 서버로 1만명의 사용자가 동시에 당첨 여부를 확인할 수 있어야 한다.
- 1명의 사용자는 평균 5장의 로또를 구매한 상태이다.

위 요구사항을 서버에서 생성되는 LottoNumber의 인스턴스의 갯수는?
1 * 5장 * 6개 숫자/1장 * 1만명 = 30만개이다.

동시에 생성되는 인스턴스 갯수가 너무 많다.
인스턴스 갯수를 줄일 수 있는 방법은?
로또 숫자 값을 LottoNumber 객체로 래핑하는 경우 매번 인스턴스가 생성되기 때문에 인스턴스의 갯수가 너무 많아져 성능이 떨어질 수 있다.
LottoNumber 인스턴스를 생성한 후 재사용할 수 있도록 구현한다.

힌트 : Map과 같은 곳에 인스턴스를 생성한 후 재사용하는 방법을 찾아본다.

javajigi commented on src/main/java/lotto/domain/LottoResult.java:
```diff
@@ -1,29 +1,28 @@
-            throw new IllegalArgumentException("로또 등수는 1개 이상이어야 합니다.");
-        }
+    public void updateRank(LottoRank rank) {
+        rankCounts.merge(rank, 1L, Long::sum);
```
👍

javajigi commented on src/main/java/lotto/domain/WinningLotto.java:
```diff
@@ -2,17 +2,31 @@
 
-    public WinningLotto(List<LottoNumber> numbers) {
-        this(new Lotto(numbers));
+    public WinningLotto(List<LottoNumber> numbers, String bonus) {
+        this(new Lotto(numbers), new LottoNumber(bonus));
+    }
```
```suggestion
    public WinningLotto(List<LottoNumber> numbers, String bonus) {
        this(new Lotto(numbers), new LottoNumber(bonus));
    }
    
    public WinningLotto(List<LottoNumber> numbers, int bonus) {
        this(new Lotto(numbers), new LottoNumber(bonus));
    }
```
int 부 생성자도 추가하는 것은?

javajigi commented on src/test/java/lotto/domain/WinningLottoTest.java:
```diff
@@ -0,0 +1,24 @@
+
+    @Test
+    void 생성자_정상입력_생성성공() {
+        assertThatCode(() -> new WinningLotto(new Lotto(1, 2, 3, 4, 5, 6), new LottoNumber(7)))
```
```suggestion
        assertThatCode(() -> new WinningLotto(new Lotto(1, 2, 3, 4, 5, 6), 7))
```
int 부 생성자 추가해 위와 같이 구현 가능하지 않을까?

javajigi reviewed (changes requested):
2단계 객체 설계가 너무 깔끔하다보니 3단계 변경이 정말 적었네요. 💯 
스트랭귤러 패턴을 이미 실무에서도 사용하고 있었다니 👍 
바로 머지해도 되지만 불변 객체로 구현할 때 발생하는 성능 이슈에 대해 연습해 봤으면 하는 바람으로 피드백 남겨봤어요.

ghtjr410 committed:
feat: WinningLotto 보너스 번호 int 입력 부생성자 추가

WinningLotto.java
- LottoNumber 대신 int로 보너스 번호를 받을 수 있는 생성자 추가

WinningLottoTest.java
- 새 int 기반 생성자 사용하도록 테스트 수정

ghtjr410 committed:
perf: LottoNumber 캐싱 적용 및 정적 팩토리 추가

LottoNumber.java
- 1~45 범위 LottoNumber 인스턴스 캐시 도입
- of(int), of(String) 정적 팩토리 메서드 추가
- 범위 검증 로직 정적 메서드로 분리
- isValidRange → isInvalidRange로 단순화

LottoNumberTest.java
- of 메서드 정상/예외 테스트 추가

ghtjr410 committed:
refactor: LottoNumber 정적 팩토리 및 캐싱 적용에 따른 전반적 코드 정리

Lotto.java
- 번호 변환을 LottoNumber::of 기반으로 변경

LottoNumber.java
- record → class로 변경
- 1~45 번호 캐싱 및 정적 팩토리 of(int, String) 적용
- 생성자 제거 및 equals/hashCode 직접 구현

WinningLotto.java
- 보너스 번호 생성 시 LottoNumber.of 사용

AutoBasedLottoGenerator.java
- 랜덤 풀 생성 시 LottoNumber::of 사용

LottoNumberParser.java
- 파싱 결과를 LottoNumber.of로 생성

LottoNumberTest.java
- 생성자 관련 테스트 제거
- 동일 값 캐시 테스트 추가

LottoNumberParserTest.java
- LottoNumber 생성 부분을 LottoNumber.of로 변경

ghtjr410 committed:
docs: LottoNumber 캐싱 기능 체크 항목 추가

03-lotto-rank-second.md
- LottoNumber 동일 값 캐싱 처리 여부 항목 추가

ghtjr410 commented:
# 피드백 반영
LottoNumber 캐싱을 적용했습니다.

처음에는 "캐싱해두면 몇 장을 사도 문제없겠다"는 접근으로 AutoBasedLottoGenerator에서 캐싱하면 충분하다고 생각했습니다.

하지만 pobi님이 제시해주신 "1만명의 사용자가 동시에 당첨 여부를 확인"하는 상황을 고민해보니, 기존 방식은 인스턴스별 캐싱이라 스레드마다 45개씩 생성되는 문제가 있었습니다.

따라서 LottoNumber 클래스 레벨에서 static 캐싱을 적용하여 모든 스레드가 45개 인스턴스를 공유하도록 변경했습니다.
***
이번에도 리뷰 잘 부탁드립니다!

javajigi commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -1,16 +1,56 @@
+    private final int value;
+
+    static {
+        CACHE = IntStream.rangeClosed(MIN, MAX).boxed().collect(Collectors.toMap(i -> i, LottoNumber::new));
```
👍

javajigi commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -1,16 +1,56 @@
+
+    public static LottoNumber of(int input) {
+        validateRange(input);
+        return CACHE.get(input);
+    }
 
-    public LottoNumber {
-        if (!isValidRange(value)) {
+    public static LottoNumber of(String input) {
```
정적 팩터리 메서드로 구현 👍

javajigi reviewed (approved):
LottoNumber에 대한 캐싱 굿 👍 
생성자를 private으로 구현함으로써 캐싱을 사용할 수 밖에 없도록 강제한 것도 👍 
2단계 -> 3단계도 스트랭귤러 패턴 적용하면서 구현했는데요.
4단계 도전도 같은 방식으로 접근해볼 것을 추천합니다.
